SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp

```diff
@@ -0,0 +1,153 @@
+using namespace esimd_test::api::functional;
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for initializer test case
+struct initializer {
+  static std::string get_description() { return "initializer"; }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
+    source_simd.copy_from(ref_data);
+    simd<DataT, NumElems> simd_by_init = simd<DataT, NumElems>(source_simd);
+    return simd_by_init;
+  }
+};
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for variable declaration test case
+struct var_declaration {
+  static std::string get_description() { return "variable declaration"; }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
+    source_simd.copy_from(ref_data);
+    simd<DataT, NumElems> simd_by_var_decl{source_simd};
+    return simd_by_var_decl;
+  }
+};
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for rvalue in an expression test case
+struct rval_in_expression {
+  static std::string get_description() { return "rvalue in an expression"; }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
+    source_simd.copy_from(ref_data);
+    simd<DataT, NumElems> simd_by_rval;
+    simd_by_rval = simd<DataT, NumElems>(source_simd);
+    return simd_by_rval;
+  }
+};
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for const reference test case
+struct const_ref {
+  static std::string get_description() { return "const reference"; }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems>
+  call_simd_by_const_ref(const simd<DataT, NumElems> &simd_by_const_ref) {
+    return simd_by_const_ref;
+  }
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
+    source_simd.copy_from(ref_data);
+    return call_simd_by_const_ref<DataT, NumElems>(
+        simd<DataT, NumElems>(source_simd));
+  }
+};
+
+// The main test routine.
+// Use struct with overloaded call operator to have able for iterating over
+// pre-determined datatypes.
+template <typename DataT, int NumElems, typename TestCaseT> struct test {
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+    bool passed{true};
+
+    vector_with_allocator<DataT> ref_data{allocator_type<DataT>{queue}};
+    generate_ref_data<DataT, NumElems>(ref_data);
+
+    auto result_data = call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
```

Comment:
```suggestion
    const auto result_data = call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
```

Reply:
I've done it in this commit: 92b81b151170f0570d3d71b4ec9c78c211298b9c